### DISCUSSION TITLE
Does xrdp support the synchronize event?

### DISCUSSION THREAD

**Discussion by Kreijstal** [neutrinolabs/xrdp#2949] [Q&A]
I am experienced a weird behaviour with both wfreerdp and sdl-freerdp on windows 8 and 10 clients, where sometimes mouse doesn't "synchronize" for example I do a drag action, and move the mouse outside the window, release the button, and click on the client again, and using cinnamon/unity/kde plasma, firefox tends to grab the mouse, I am unable to click other windows, I can't click the minimize button, etc. The interesting behaviour is that this persist even if I switch client, for example when using guacd this bug never occurs, however when I manage to reproduce the bug, I disconnect, then connect to guacd the bug seems to persist, and the only thing that seems to fix it, is to kill the enviroment. `pkill cinnamon`, Simply 'resetting' it doesn't work most of the time. I've communicated this to freerdp team, and this is an extract of the coversation:

```
<kreijstal> So not sure if anyone has had this problem with wfreerdp or sdl-freerdp from win8/10 controlling xrdp, but sometimes mouse clicks tend to go for too long as if the mouse never registered the up clicks, not sure if it is a freerdp or xrdp bug. But I have to restart cinnamon or kill it so it starts working again.
<kreijstal> Something that tends to happen is that firefox likes to 'grab' the mouse, and I can't click on the minimize,maximize or DE buttons.
<kreijstal> This happens with wfreerdp and sdl-freerdp on windows 8 and 10 controlling linux mint 
<kreijstal> I also know this bug does not happen when I am using guacd, but it still could be a xrdp bug, reporting here in case anyone has ideas, even if just to submit a better bug report.
<kreijstal> I also considered it being a cinnamon bug, but it also happens with KDE guests.
<kreijstal> And now that I remember with ubunu's  unity DE too
<akallabeth[m]> kreijstal: you can do a build with `WITH_DEBUG_SDL_EVENTS=ON` for `sdl-freerdp`, then you get `SDL_MOUSEBUTTONDOWN` and `SDL_MOUSEBUTTONUP` log messages (along with lots of others) 
<akallabeth[m]> if you experience the bug and the last event was a `SDL_MOUSEBUTTONUP` then it is most likely a bug at the remote end. 
<akallabeth[m]> <shrait[m]> "In wfreeRDP version 3.2, are UDP..." <- https://github.com/FreeRDP/FreeRDP/issues/4978 is still not closed (as in not implemented)...
<akallabeth[m]> this should answer all questions ;)
<kreijstal> akallabeth[m]: Not sure if it is a bug on xrdp, because when the bug occurs the bug remains on the DE even after I switch rdp clients, say the bug occurs on sdl-freerdp, then I go to guacd, and the bug persists even though it would never occur with just guacd. The DE buttons stop responding, I can't minimize windows or click buttons, and sometimes it appears as if the click is being pressed always. I would say it is xrdp bug but it only happens with 
<kreijstal> {w,sdl-}freerdp
<kreijstal> maybe a double bug
<kreijstal> should I report both sides
<akallabeth[m]> kreijstal: just means that the state is kept, so once it is broken you need to reset it
<kreijstal> I've found I can reset it by: restarting cinnamon (sometimes it works) but even then not always and I have to resort to `pkill cinnamon`. Is there a better way to reset this state you are talking about.
<akallabeth[m]> yes. still would be interesting if 
<akallabeth[m]> 1. the client does not send the button up
<akallabeth[m]> 2. the RDP server just does not pass it properly
<pnowack[m]> <kreijstal> "akallabeth: Not sure if it is..." <- Does xrdp handle the `Synchronize` event? The server side has to track pressed buttons and release them, when that event is received. The server side also needs to release all keys, when the client is gone, as otherwise there won't ever be key_release events for the pressed keys.
```

Of course I know this is too little information to go on I have yet to compile freerdp using the events, and I don't have enough information for a complete bug report, but I wanted to ask if someone knows or has experienced something similar.

**Comment by matt335672** [comment]
The sync event that @akallabeth mentions is essentially documented here :-

https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/6c5d0ef9-4653-4d69-9ba9-09ba3acd660f

It's very useful for 'sticky' keys like caps lock and scroll lock, as for these we need to know the state of the keys to map keypresses to input data. For keys other than the sticky ones the down/up events aren't important to the server, and we just pass them straight through without storing their state.

So, yes we do process the synchronise event, but we don't specifically apply it to mouse buttons.
